server/src/main/java/com/hedera/block/server/persistence/cache/LRUCache.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.block.server.persistence.cache;
+
+import com.hedera.block.protos.BlockStreamServiceGrpcProto;
+import jakarta.inject.Singleton;
+
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * An LRU cache implementation which uses a custom LinkedHashMap to store blocks in memory and
+ * evict the least recently used block when the cache size exceeds a specified limit.
+ */
+@Singleton
+public class LRUCache implements BlockCache<BlockStreamServiceGrpcProto.Block> {
+
+    private final Map<Long, BlockStreamServiceGrpcProto.Block> cache;
+
+    /**
+     * Constructor for the LRUCache class.
+     *
+     * @param maxEntries the maximum number of entries in the cache
+     */
+    public LRUCache(final long maxEntries) {
+        final System.Logger LOGGER = System.getLogger(getClass().getName());
+        LOGGER.log(System.Logger.Level.INFO, "Creating LRUCache with maxEntries: " + maxEntries);
+
+        cache = Collections.synchronizedMap(new BNLinkedHashMap<>(maxEntries));
+    }
+
+    /**
+     * Inserts a block into the cache.
+     *
+     * @param block the block to insert
+     * @return the id of the block
+     */
+    @Override
+    public Long insert(final BlockStreamServiceGrpcProto.Block block) {
+        final long id = block.getId();
+        cache.putIfAbsent(id, block);
```

Comment:
This prevents updating a block, which is very likely to be necessary.  Also, note that block _items_ do not have numeric identifiers (and storing whole blocks is excessive, a single block may be a gigabyte or more and contain tens or hundreds of thousands of block items).

Block Node very likely will not, in the ultimate, be _able_ to accept whole blocks and operate on them efficiently.  The much more likely paradigm is to process everything in a streaming fashion, one block _item_ at a time, retaining only the minimum necessary for block validation (e.g. hash values) in memory with everything else consigned immediately to persistent storage.

Responding to queries should be similar, stream items from storage to the querent with only one or two items held in memory at any given moment.

Reply:
Thanks for the additional context around size and processing constraints around Blocks and BlockItems.  The type "Block" here is just meant as a placeholder for what will come later via PBJ (blockstream.proto is a toy implementation to facilitate development).  I'm open to renaming the types in a subsequent PR to more closely model the real thing.  Regarding numeric identifiers, my recollection from stand-up discussions with Jasper is that they were considering numbering BlockItems so that the Block Node server would know the ordering to reassemble them into a block for storage.  Maybe we need to have a further discussion around what this will look like.

I've removed the cache.  We can revisit the idea later as needed.